winit-wayland/src/window/state.rs

```diff
@@ -128,47 +126,64 @@ pub struct WindowState {
 
-    /// Whether the CSD fail to create, so we don't try to create them on each iteration.
-    csd_fails: bool,
-
     /// Whether we should decorate the frame.
     decorate: bool,
 
-    /// Min size.
-    min_surface_size: LogicalSize<u32>,
-    max_surface_size: Option<LogicalSize<u32>>,
-
-    /// The size of the window when no states were applied to it. The primary use for it
-    /// is to fallback to original window size, before it was maximized, if the compositor
-    /// sends `None` for the new size in the configure.
-    stateless_size: LogicalSize<u32>,
-
     /// Initial window size provided by the user. Removed on the first
     /// configure.
     initial_size: Option<Size>,
 
     /// The state of the frame callback.
     frame_callback_state: FrameCallbackState,
 
-    viewport: Option<WpViewport>,
-    fractional_scale: Option<WpFractionalScaleV1>,
     blur: Option<OrgKdeKwinBlur>,
     blur_manager: Option<KWinBlurManager>,
 
     /// Whether the client side decorations have pending move operations.
     ///
     /// The value is the serial of the event triggered moved.
     has_pending_move: Option<u32>,
+}
+
+#[derive(Debug)]
+enum ShellSpecificState {
+    Xdg {
+        /// The underlying SCTK window.
+        window: Window,
+
+        /// The last received configure.
+        last_configure: Option<WindowConfigure>,
+
+        /// Whether the frame is resizable.
+        resizable: bool,
 
-    /// The underlying SCTK window.
-    pub window: Window,
+        // NOTE: The spec says that destroying parent(`window` in our case), will unmap the
+        // subsurfaces. Thus to achieve atomic unmap of the client, drop the decorations
+        // frame after the `window` is dropped. To achieve that we rely on rust's struct
+        // field drop order guarantees.
+        /// The window frame, which is created from the configure request.
+        frame: Box<Option<WinitFrame>>,
 
-    // NOTE: The spec says that destroying parent(`window` in our case), will unmap the
-    // subsurfaces. Thus to achieve atomic unmap of the client, drop the decorations
-    // frame after the `window` is dropped. To achieve that we rely on rust's struct
-    // field drop order guarantees.
-    /// The window frame, which is created from the configure request.
-    frame: Option<WinitFrame>,
+        /// Whether the CSD fail to create, so we don't try to create them on each iteration.
+        csd_fails: bool,
+
+        /// The size of the window when no states were applied to it. The primary use for it
+        /// is to fallback to original window size, before it was maximized, if the compositor
+        /// sends `None` for the new size in the configure.
+        stateless_size: LogicalSize<u32>,
+
+        /// Min size.
+        min_surface_size: LogicalSize<u32>,
+        max_surface_size: Option<LogicalSize<u32>>,
+
+        viewport: Option<WpViewport>,
+        fractional_scale: Option<WpFractionalScaleV1>,
```

Comment:
Is fractional scaling not supported for layershell?

Reply:
Shouldn't this field be in the shared part of the window state and not in xdg specific one? unless I am missing something

Reply:
I think it's a good idea, I will make the change

Reply:
Done @Exidex, thanks for your suggestion 😃

Reply:
@SergioRibera another thing I noticed is that fractional scaling relies on wp_viewport to work correctly. I fixed it in my fork, which is based on 0.30.8, so it may not translate well, but you can use it as reference : https://github.com/project-gauntlet/winit/commit/76548141d9975bfd84c2031f27cda19224fa6bcf

Reply:
@Exidex Thank you very much for what you did. I uploaded a commit after spending some time fixing what you recommended 🎉